### ISSUE TITLE
Run as non-root user

### ISSUE THREAD

**Issue by jcatana** [kubeflow/mpi-operator#301]
Are there any plans for an implementation of mpi-operator that can function as user other than root?

**Comment by terrytangyuan** [comment]
No plan on this yet. If you have any suggestions on how to approach this and would like to contribute, please let us know here!

**Comment by jcatana** [comment]
From what I can tell, it seems the only part that requires root is for updating the /etc/hosts file. Please correct me if I'm wrong.

I've been thinking on a few ideas of how to approach it. If this is the only reason, the solution would be finding an alternate method for making the job members resolvable. 

What I've attempted is to create a headless service for the stateful set [as described here](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#stable-network-id). I could imagine using this technique could potentially cause a race condition in some clusters where DNS is not yet resolvable when mpirun execution is attempted, but this could be scripted around. Something that checks name resolution prior to issuing the exec and launching mpirun. I haven't personally seen any issue with not having all members being immediately resolvable.

I've been hacking around with the code and been able to add the creation of a headless service to the stateful set creation, it seems to work, but MPI fails to execute because the launcher, in it's current state, is a job member and also needs to be resolvable.

I can think of 2 different solutions for making the launcher resolvable.

1. Use kubernetes built in [HostAlias](https://kubernetes.io/docs/concepts/services-networking/add-entries-to-pod-etc-hosts-with-host-aliases/) I've attempted to add this to the stateful set, but it appears the statefulset is defined by the controller at submission, before the launcher job actually begins running, so the API query to return the launcher's hostname and IP to add to the stateful set are empty.

2. Don't include the launcher as part of the MPI job. Exec into the rank0 worker and execute mpirun from there. I haven't fully fleshed out this idea and having an inter-pod exec controlling the job (that could last hundreds of hours) might be doomed.

Potentially an alternate approach might be an init container that creates a world writable hosts file in an empty dir volume that overwrites the main container /etc/hosts

I'm sure there are others that may have much more elegant solutions for how to approach this. I appreciate you taking the time to entertain my ideas and request.

**Comment by AssafKatz3** [comment]
Hi,
I have stupid question - why does this operator need to change the /etc/hosts in first place? This file is managed by k8s...
Thanks

**Comment by jcatana** [comment]
Having all MPI worker names resolvable is an MPI requirement.

**Comment by AssafKatz3** [comment]
@jcatana This is clear, but what the information that service name doesn't contain?

**Comment by jcatana** [comment]
That is sort of the issue. Currently no service is created for the stateful set. To further complicate the issue, kubernetes services cannot address kubernetes Jobs. This is why I talk about utilizing the hostAlias parameter in the statefulSet definition.

A kubernetes MPIJob is comprised of two objects:
Launcher -> Kind: Job -> this executes the mpirun command.
Workers -> Kind: StatefulSet -> these are the targets of mpirun command.

Neither of these Kind create integrated kubernetes DNS entries, which is why, I'm guessing, MPI operator modifies the local /etc/hosts files in the pods. Modification of this file requires root (usually...). All participating entities in the MPI job must be resolvable, workers and launcher.

My proposal is to use built in kubernetes methods of making the objects resolvable so modification of /etc/hosts is not necessary and therefore root would not be required.

**Comment by AssafKatz3** [comment]
@jcatana Kubernetes DNS entries are created by services, which the most common why to define them is by selector to PODs. Actually, it doesn't matter if the POD was created by deployment, stateful set, job or directly. The only possible problem that I see with this response time until the PODs of job will be added to endpoint (or endpoint slice) object that service relies on.

**Comment by AssafKatz3** [comment]
@jcatana I looked at code and there are two references to /etc/hosts:

1. [generateHosts](https://github.com/kubeflow/mpi-operator/blob/master/pkg/controllers/kubectl_delivery/controller.go#L162) that add to it, this code can be replaced by generate service, either by set selector or more directly by creating endpoint object and then service on top of it
2. [newConfigMap](https://github.com/kubeflow/mpi-operator/blob/master/pkg/controllers/v1alpha2/mpi_job_controller.go#960) the code copy the information from file by kubectl cp, bad idea anyway (this command rely on a lot of other stuff - existence of tar, writable root, etc.) and can be replaced by putting the services' names.

Note that newConfigMap is for v1alpha2 so I am not sure that it is really need to be fixed it, and therefore generateHosts isn't needed (or it will be possible to turn it off by some flag)

**Comment by jcatana** [comment]
Yes, this solution would also work, but it has the same issue as hostAlias. You do not know the podSelector label until after the Job's pod is created much like for hostAlias you do not know the pod's IP until after it is created.

Here is my code that will create the headless service for stateful set https://github.com/jcatana/mpi-operator/ IIt also modifies the sh script that is run not to attempt to modify the hosts file.
've only been playing with V1alpha2.  I'd be happy to collaborate on the last piece of the puzzle.

**Comment by kolasaniv1996** [comment]
is there any further update for supporting non root users?

**Comment by alculquicondor** [comment]
The example for v2beta1 is actually written for non-root
https://github.com/kubeflow/mpi-operator/blob/master/examples/v2beta1/pi/pi.yaml

/close

**Comment by kolasaniv1996** [comment]
@alculquicondor  if you run with any other user it will not work.

**Comment by alculquicondor** [comment]
You might have to build your own image to change the user id https://github.com/kubeflow/mpi-operator/blob/52cda2c7e85ac22284ea23e1d905c4e2eaefdc11/build/base/Dockerfile#L27-L31

Or somehow parameterize it.

**Comment by kolasaniv1996** [comment]
after building the docker image, do i have to re install the mpi operator ? if yes can you guide ?

**Comment by alculquicondor** [comment]
no, this is the docker image for your job.

The mpi-operator image doesn't need to be changed or rebuilt.

**Comment by kolasaniv1996** [comment]
I have successfully tested the updated image using a user with UID 1001. The solution involves running the MPI operator with this specific user ID (UID 1001). However, if any individual user wants to use the operator, they will need to build their own Docker image with their respective user ID (UID). This ensures that the MPI operator can run without issues :-

Dockerfile Customization:

https://github.com/kolasaniv1996/MPI-RUNAI/blob/main/Dockerfile

Key Points:
The UID in the Dockerfile must be replaced with your own UID for the mpiuser.
The user will be created as mpiuser with your custom UID to avoid permission issues.
The sshd_config file is configured to allow SSH operations without root access

ssh_config-

https://github.com/kolasaniv1996/MPI-RUNAI/blob/main/sshd_config

MPI Job YAML:
Once you build and push your custom Docker image, you will need to adjust the YAML file to specify the user ID accordingly.



https://github.com/kolasaniv1996/MPI-RUNAI/blob/main/mpi-nonroot.yaml
